### DISCUSSION TITLE
Implement custom driver in  rocket_db_pools

### DISCUSSION THREAD

**Discussion by d-ufrik** [rwf2/Rocket#2544] [Questions] [answered]
Hello.

As a PoC i would like to create a driver for a Oracle DB Free using [rocket_db_pools]. I've read the documentation several times and i can't find a way to make it work properly. 

I'm not asking for specific code or implementation, if you please point me in the right direction to understand how to start as if it were a general driver , would really appreciate that.

I've tried deriving directly from trait rocket_db_pools::Database just to see if i could understand the errors and fix them, evidently, i failed miserably.

Could you please point me into the right:

```
struct Oracle(oracle::pool::Pool);


#[rocket::async_trait]
impl Pool for Oracle{
    type Connection = oracle::Connection;

   type Error = oracle::Error;

    async fn init(figment: &Figment) -> Result<Self, Self::Error> {
        //todo!("initialize and return an instance of the pool");
        let user = "HR";
        let password = "password1";
        let connect_string = "//127.0.0.1:5121/FREEPDB1";
        let pool = PoolBuilder::new(user, password, connect_string)
            .max_connections(20)
            .build()?;
            Ok(Oracle(pool))
    }


    async fn get(&self) -> Result<Self::Connection, Self::Error> {
        //todo!("fetch one connection from the pool");
        Ok(self.get().await?)
    }

    async fn close(&self) {
        todo!("gracefully shutdown connection pool");
    }
}
```

**Comment by SergioBenitez** [comment, marked as answer]
Have you taken a close look at the [implementing](https://api.rocket.rs/v0.5-rc/rocket_db_pools/trait.Pool.html#implementing) part of the docs? It should be sufficient to guide you to a working implementation. Once you implement `Pool` for some type, you can use it as the target of the `database` attribute. 

Here are some more guidelines:

  * `rocket_db_pools` exists to integrate _existing_ `async` pools with Rocket applications. As such, you'll need an existing `async` database pool for your database of choice. In your case, this means you need to find or write an `async` connection pool for Oracle. If the pool isn't `async` (and instead `sync`chronous), you'll need to use `rocket_sync_db_pools` instead.
  * You implement `Pool` for the pool type. If you're using a pool from another crate, you'll need to wrap the type to get around Rust's orphan impl rules, like this: `struct OraclePool(external::OraclePool)`. You'd then implement everything for that type.

This all looks like:

```rust
impl Pool for OraclePool { ... }

#[database("oracle")]
struct MyDb(OraclePool);

#[get("/")]
fn index(db: Connection<MyDb>) {
    use_connection(&*db);
}
```

**Reply by d-ufrik** [reply]
Thank you very much, Sergio.
